### DISCUSSION TITLE
LDAP Issues

### DISCUSSION THREAD

**Discussion by Raven942** [kimai/kimai#3056] [Questions & Answers] [answered]
Hello,

I am having issues getting LDAP to function. currently i have it disabled. ive attached my local.yaml file below. Any idea where im going wrong?

kimai:
    user:
      registration: false
      password_reset: false
    permissions:
      roles:
          ROLE_USER: ['!password_own_profile']
          ROLE_TEAMLEAD: ['!password_own_profile']
          ROLE_ADMIN: ['!password_own_profile']
    ldap:
        activate: false
        connection:
            host: ldap://X.X.X.X
            port: 389
            username: adminaccount@domain.local
            password: password
            accountDomainName: domain.local
            accountDomainNameShort: domain
            accountFilterFormat: (&(objectClass=person)(sAMAccountName=%s))
        user:
            baseDn: dc=domain,dc=local
            filter: (&(objectClass=person))
            usernameAttribute: samaccountname
            attributesFilter: (objectClass=person)
            attributes:
                - { ldap_attr: mail, user_method: setEmail }
                - { ldap_attr: displayname, user_method: setAlias }
                - { ldap_attr: samaccountname, user_method: setUsername }
        role:
            baseDn: dc=domain,dc=local
            filter: (&(objectClass=group))
            groups:
                - { ldap_value: PW Foremans, role: ROLE_TEAMLEAD }
                - { ldap_value: Domain Admins, role: ROLE_SUPER_ADMIN }
                - { ldap_value: PW Laborers, role: ROLE_USER }
                - { ldap_value: PW Supervisors, role: ROLE_ADMIN }

**Comment by kevinpapst** [comment]
> I am having issues

This needs to be described in detail with logs

**Reply by Raven942** [reply]
the log is completely clean, there is nothing in var/log/prod.log

**Reply by Raven942** [reply]
i activated it an attempted login, i get nothing, not even in the logs.

**Comment by schulzc21** [comment]
Hello,

same for me here.
Configured LDAP as described in https://www.kimai.org/documentation/ldap.html, trying to log in with a test user account I only get

"Die Authentifikation konnte wegen eines Systemproblems nicht bearbeitet werden."

-> Nothing in var/log/prod.log

Do you have any hint where to look for errors?

Sorry and thank you for your remarkable work!

Christian

**Reply by Raven942** [reply]
I found this, which this person has at least something working. Still a bit unclear.

https://github.com/kevinpapst/kimai2/issues/875

**Comment by kevinpapst** [comment, marked as answer]
Everyone: there have been enough issues and discussions in the past. The issues are closed, so check that when using the search.
Read through the issues, activate debugging, activate logging on your LDAP server to see the queries.
Activate DEV mode to see more logs. 
And if you want paid support, please sent me an email.

**Reply by Raven942** [reply]
I have searched, followed the recommended setup for active directory, I’ve just now learned I may be missing some items in my local.yaml with security and firewalls.

Im still unclear if I need to have the users already added into Kimaj or if it pulls that person from successful logins.

The suggested setup for AD on the documentation shows one way to use the username, but the person in the following thread has it set up completely differently.

https://github.com/kevinpapst/kimai2/issues/875

just some confusion is all.

**Reply by kevinpapst** [reply]
Posting questions while the answer is actually the first real sentence in the documentation:

> LDAP users will be imported during the first login and their attributes and groups updated on each following login.

**Reply by Raven942** [reply]
Yes that makes sense, and why my Users are not added in the first place, but because it has never worked, typically the default action is to attempt other avenues of troubleshooting. The article referenced, Someone specifically asks the guy if he had users added in there that we’re trying to login. This creates the question in mind.

Additionally, outside of the documentation, these discussions are the only thing on the internet people can reference. My attempt to be complete and provide answers in the discussions with references to others is a way to try and help people save time. There are “many others who have this working fine” and “don’t have configuration issues” indicates that it most likely is something dumb, like a simple missing line in the local.yaml or something with the wrong parameters. That’s why posting the config in hopes of someone saying “oh, you don’t have xxxxx in there” or “it needs to be this format” isn’t really a far fetched idea.

while we all appreciate your hard work and dedication, typically we are seeking answers from our peers on here for solutions they have come up. Some of them have a native environment in Linux based OS, mine is not. You answering is amazing, considering the workload and you are the developer. I commend you for it, but I think the hopes are the community will answer.

there isn’t a tutorial out that clearly defines settings for different distorts either. That’s not your responsibility, really none of this is. But it all leads to some of these growing pains of figuring things out.

I can’t thank you enough for creating this software, for the time you sacrifice to work on it, for the responses you give, or the countless hours you spend attempting to improve things. Without you we wouldn’t be able to have this discussion. I just believe we were all hoping for some community engagement from others.

**Reply by schulzc21** [reply]
...got it up and running thanks to your hints, Kevin - for Drozzo: here is my config which works well for a login. Will have a look at group/role mapping now.

kimai:
    user:
        registration: false
        password_reset: false
    ldap:
        activate: true
        connection:
            host: x.y.local
            username: name
            password: password
            bindRequiresDn: false
            accountDomainName: x.y
            accountDomainNameShort: X
            accountFilterFormat: (&(objectClass=Person)(sAMAccountName=%s))
        user:
            baseDn: dc=x,dc=y
            filter: (&(objectClass=Person))
            usernameAttribute: samaccountname
            attributesFilter: (objectClass=Person)
            attributes:
                - { ldap_attr: mail, user_method: setEmail }
                - { ldap_attr: displayname, user_method: setAlias }
                - { ldap_attr: samaccountname,  user_method: setUsername }
        role:
            baseDn: ou=gruppen,dc=x,dc=y
            filter: (&(objectClass=group))
            groups:
                - { ldap_value: Leads, role: ROLE_TEAMLEAD }
                - { ldap_value: Sysadmins, role: ROLE_SUPER_ADMIN }
                - { ldap_value: Users, role: ROLE_USER }

Regards,
Christian

**Reply by Raven942** [reply]
Christian, thank you.

what was the issue originally? Also where you have username is it literally just the username with no domain attributes attached? (Ie admin, and not admin@domain or domain/admin)

**Reply by Raven942** [reply]
Also, did you have anything in there about firewall or security? I thought all that was handled in the security.yaml so I didn’t place it in mine

**Reply by schulzc21** [reply]
...the issue was - trying to log in with a test user account - I only got

"Die Authentifikation konnte wegen eines Systemproblems nicht bearbeitet werden."

while no user was created in kimai.
My posting above contains the whole config, so no settings about firewall/security etc.
The username has no domain attributes attached, in your example "admin" without the quotes.

In my case, the baseDn setting did not work out when I used 

ou=users,dc=x,dc=y

in spite of the fact that my test user resides in the OU users.
Using only

dc=x,dc=y

as baseDN did the trick, and I still don't understand why.

**Reply by Raven942** [reply]
My guess is the base en has to be the root base to recognize the structure. It’s the filtering following it that regulates what’s actually being pulled. I have a Cisco forewall that is properly performing AD sync in that manner. I’ll try to pull those configs Monday to help compare. 

You having the login as only the name is helpful information, as the documentation doesn’t show that. It’s shows it with @domain.local following.

**Reply by Raven942** [reply]
Update: Changing my configuration to the above syntax and reloading cache finally got this to work. The documentation on getting AD to work should reflect these settings. i went ahead and added the port as well for those who would have to use secure LDAP.

kimai:
    user:
      registration: false
      password_reset: false
    permissions:
      roles:
          ROLE_USER: ['!password_own_profile']
          ROLE_TEAMLEAD: ['!password_own_profile']
          ROLE_ADMIN: ['!password_own_profile']
    ldap:
        activate: true
        connection:
            host: name.domain.local
            port: 389
            username: username (ONLY NAME)
            password: password
            bindRequiresDn: false
            accountDomainName: domain.local
            accountDomainNameShort: domain
            accountFilterFormat: (&(objectClass=person)(sAMAccountName=%s))
        user:
            baseDn: dc=domain,dc=local
            filter: (&(objectClass=person))
            usernameAttribute: samaccountname
            attributesFilter: (objectClass=person)
            attributes:
                - { ldap_attr: mail, user_method: setEmail }
                - { ldap_attr: displayname, user_method: setAlias }
                - { ldap_attr: samaccountname, user_method: setUsername }
        role:
            baseDn: dc=domain,dc=local
            filter: (&(objectClass=group))
            groups:
                - { ldap_value: Foremans, role: ROLE_TEAMLEAD }
                - { ldap_value: Domain Admins, role: ROLE_SUPER_ADMIN }
                - { ldap_value: Laborers, role: ROLE_USER }
                - { ldap_value: Supervisors, role: ROLE_ADMIN }

**Comment by kevinpapst** [comment]
And to both of you and everyone else having a problem "does not work" and "there is no log" => you are IT as well, could you help someone with that description?

You might have to switch to `APP_ENV=dev` and run `composer install` to get more logs.
If you want help, you really have to share details what you tried, what did not work, how is your LDAP structure....

Everyone setting up LDAP is doing that for a company. Why not asking your boss for an hour of consultancy by the guy that created the entire software for free. My time is limited. Maybe someone else reads this here, who is actually using an LDAP and wants to help.
